packages/go/analysis/ad/adcs.go

```diff
@@ -161,6 +165,160 @@ func PostADCSESC3(ctx context.Context, tx graph.Transaction, outC chan<- analysi
 	return nil
 }
 
+func principalControlsCertTemplate(principalID uint32, certTemplate *graph.Node, groupExpansions impact.PathAggregator, cache ADCSCache) bool {
+	controllers := cache.CertTemplateControllers[certTemplate.ID]
+	for _, controller := range controllers {
+		if principalID == controller.ID.Uint32() {
+			return true
+		}
+		//Since CertTemplateControllers only contains first degree controllers, if the principal is a group we do a group expansion to check for certTemplate control through nested members
+		if controller.Kinds.ContainsOneOf(ad.Group) {
+			expanded := groupExpansions.Cardinality(controller.ID.Uint32()).(cardinality.Duplex[uint32])
+			if expanded.Contains(principalID) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func checkEmailValidity(node *graph.Node, validCertTemplates []*graph.Node, groupExpansions impact.PathAggregator, cache ADCSCache) cardinality.Duplex[uint32] {
+	var (
+		results   = cardinality.NewBitmap32()
+		email, _  = node.Properties.Get(common.Email.String()).String()
+		exception = false
+	)
+	if email == "" {
+		for _, certTemplate := range validCertTemplates {
+			if principalControlsCertTemplate(node.ID.Uint32(), certTemplate, groupExpansions, cache) {
```

Comment:
You might want to consider caching the result of this somewhere so that we don't keep looking up the same principal over and over again. That being said, I _think_ the cost is relatively cheap so it might not be a big deal

Reply:
Tacking this property onto the adcscache made sense to me so that is what I did. Let me know if you think it should be scoped more narrowly and/or defined elsewhere.